File: c-plus-plus/primes/Simple.cpp
#include <iostream>
#include <chrono>
#include <cmath>

static const auto NUMBER = 500000;

auto getLastPrime(int count) -> int
{
    auto lastPrime = 0;

    for (auto i = 3; i <= count; i += 2) {
        bool isPrime = true;
        int limit = static_cast<int>(std::sqrt(i));

        for (auto j = 3; j <= limit; j += 2) {
            if (i % j == 0) {
                isPrime = false;
                break;
            }
        }

        if (isPrime) {
            lastPrime = i;
        }
    }

    return lastPrime;
}

int main() {
    const auto start_time = std::chrono::high_resolution_clock::now();

    const auto last_prime = getLastPrime(NUMBER);
    std::cout << "Last prime: " << last_prime << std::endl;
  
    const auto end_time = std::chrono::high_resolution_clock::now();
    const auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end_time - start_time).count();

    std::cout << "Execution time: " << duration << "ms" << std::endl;

    return 0;
}

File: fortran/primes/Simple.f90
program benchmark_primes
    implicit none
    integer, parameter :: NUMBER = 500000
    integer :: lastPrime
    real :: start_time, end_time, duration

    ! Get the starting time
    call cpu_time(start_time)

    ! Find the last prime number
    lastPrime = getLastPrime(NUMBER)
    print *, 'Last prime: ', lastPrime

    ! Get the ending time
    call cpu_time(end_time)

    ! Calculate the duration in milliseconds
    duration = (end_time - start_time) * 1000.0
    print '(A, F6.2, A)', 'Execution time: ', duration, 'ms'

contains

    function getLastPrime(count) result(lastPrime)
        integer, intent(in) :: count
        integer :: lastPrime
        integer :: i, j
        integer :: limit
        logical :: isPrime

        lastPrime = 0

        do i = 3, count, 2
            isPrime = .true.
            limit = int(sqrt(real(i))) ! Check up to the square root of i

            do j = 3, limit, 2
                if (mod(i, j) == 0) then
                    isPrime = .false.
                    exit
                end if
            end do

            if (isPrime) then
                lastPrime = i
            end if
        end do
    end function getLastPrime

end program benchmark_primes
